config/config.go

```diff
@@ -0,0 +1,145 @@
+package config
+
+import (
+	"fmt"
+	"github.com/fluent/fluent-bit-go/output"
+	"strconv"
+	"unsafe"
+)
+
+type PluginConfig struct {
+	BufferManagerConfig BufferConfig
+	NRClientConfig      NRClientConfig
+	ProxyConfig         ProxyConfig
+}
+
+type BufferConfig struct {
+	MaxBufferSize         int64
+	MaxRecords            int64
+	MaxTimeBetweenFlushes int64
+}
+
+type NRClientConfig struct {
+	Endpoint   string
+	ApiKey     string
+	LicenseKey string
+	UseApiKey  bool
+}
+
+type ProxyConfig struct {
+	IgnoreSystemProxy bool
+	Proxy             string
+	CABundleFile      string
+	CABundleDir       string
+	ValidateCerts     bool
+}
+
+func NewPluginConfig(ctx unsafe.Pointer) (cfg PluginConfig, err error) {
+	cfg.BufferManagerConfig, err = parseBufferConfig(ctx)
+	if err != nil {
+		return
+	}
+
+	cfg.NRClientConfig, err = parseNRClientConfig(ctx)
+	if err != nil {
+		return
+	}
+
+	cfg.ProxyConfig, err = parseProxyConfig(ctx)
+	if err != nil {
+		return
+	}
+
+	return cfg, nil
+}
+
+func parseBufferConfig(ctx unsafe.Pointer) (cfg BufferConfig, err error) {
+	maxBufferSize, err := optInt64(ctx, "maxBufferSize", 256000)
+	if err != nil {
+		return
+	}
+	cfg.MaxBufferSize = maxBufferSize
+
+	maxRecords, err := optInt64(ctx, "maxRecords", 1024)
+	if err != nil {
+		return
+	}
+	cfg.MaxRecords = maxRecords
+
+	maxTimeBetweenFlushes, err := optInt64(ctx, "maxTimeBetweenFlushes", 5000)
+	if err != nil {
+		return
+	}
+	cfg.MaxTimeBetweenFlushes = maxTimeBetweenFlushes
+
+	return cfg, err
```

Comment:
line 53 returns `nil` instead of `err` -- why the difference?

Reply:
Good spot! I was making use of a short notation (named result parameters) provided by Go, but I was inconsistent between these two methods. I fixed it so I use a consistent notation across he whole file.

Basically, I provide a name for the return parameters in the method signature: `cfg` and `err`. This allows assigning values to them as the function executes. If at any point `err` takes a value different than `nil`, I simply `return`, which is equivalent to `return cfg, err`. Note that in this case, `cfg` may have an incomplete value, but the rule is that if the returned error `err` is not `nil`, the caller of the function should ignore the `cfg` result.

Reference: https://golang.org/doc/effective_go.html#named-results